library/alloc/src/macros.rs

```diff
@@ -99,9 +108,16 @@ macro_rules! vec {
 /// format!("hello {}", "world!");
 /// format!("x = {}, y = {y}", 10, y = 30);
 /// ```
+#[allow_internal_unstable(display_type_guard)]
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
 macro_rules! format {
+    // A faster path for simple format! calls.
+    ("{}", $($arg:tt,?)+) => {{
```

Comment:
nit: does this need to deal with a full `$($arg:tt,?)+`?  Maybe it could just take `$e:expr ,?` and let all the more complicated cases still be handled by the other arms of the macro.

(I don't think there's a need to optimize `format!("{}", 1, 2, 3, 4, 5, 6)`.)

Reply:
What's worse than that: I'm not sure that even the expr version will even be truly correct.

Right now, `format!` parses named arguments, like `a = 1`. If that was parsed as an expression, then it would evaluate to `()`, and this would not pass the assertion.

```rust
let a = format!("{}", x = 1);
assert_eq!(&a, "1");
```

I think it actually needs to be `$arg:ident`, and the optimization probably wouldn't be triggered very often as a result. That's annoying.

Reply:
Oh wow, I didn't think about that kind of usage.
@notriddle Can you post or link me to some more... uh... weird examples like that? I mean, it might make me actually just close this PR but I'd like to have an opportunity to fool around with the alternate cases to see how I can handle it in the pattern matching and if it destroys the gains.

Reply:
https://doc.rust-lang.org/std/fmt/index.html#usage

Reply:
Right, I'm aware of the usual usages, but that doesn't include any examples that I can see which match the `format!("{}", x = 1);` case that I can see, I've never seen a named parameter then being passed to an "anonymous" formatting parameter.

Reply:
> but that doesn't include any examples that I can see which match the `format!("{}", x = 1);` case that I can see

That's because it's a bug and shouldn't be accepted: #45256